executing/executing.py

```diff
@@ -509,9 +530,556 @@ def visit_ClassDef(self, node):
         self.stack.pop()
 
 
+def node_and_parents(node):
+    while True:
+        yield node
+        if hasattr(node, "parent"):
+            node = node.parent
+        else:
+            break
+
+
+
+
+class PositionNodeFinder(object):
+    """
+    Mapping bytecode to ast-node based on the source positions, which where introduced in pyhon 3.11.
+    In general every ast-node can be exactly referenced by its begin/end line/col_offset, which is stored in the bytecode.
+    There are only some exceptions for methods and attributes.
+    """
+
+    types_cmp_issue_fix = (
+        ast.IfExp,
+        ast.If,
+        ast.Assert,
+        ast.While,
+    )
+    types_cmp_issue = types_cmp_issue_fix+ (
+        ast.ListComp,
+        ast.SetComp,
+        ast.DictComp,
+        ast.GeneratorExp,
+    )
+
+    last_code = None
+    last_instructions = None
+
+    def __init__(self, frame, stmts, tree, lasti, source):
+
+        # caching the last result of get_instructions is a huge speedup for searches in the same bytecode
+        # a lru_cache with a small size would be better but is not available on older python versions
```

Comment:
yes, you are right. I will move the whole node finder to its own file and include it only conditionally. This will also simplify some other parts of the code.

Reply:
Yes, putting everything in one big file was a mistake on my part that's becoming especially costly now. Splitting it up would be nice, especially if it allows using Python 3 syntax - maybe you can even find an excuse to use pattern matching. But I don't want any reorganisation that will slow this PR down.

Reply:
I also thought about pattern matching but multiple `inst_match` calls for example might get even more complicated. And yes I also want to get this done :smile:. 
Big thank you again for your review, I really like your feedback.
This is quite complicated and turned my head more than once upside down.